javascript/src/URDFClasses.js

```diff
@@ -253,26 +258,27 @@ class URDFJoint extends URDFBase {
                 this.jointValue[0] = values[0] !== null ? values[0] : this.jointValue[0];
                 this.jointValue[1] = values[1] !== null ? values[1] : this.jointValue[1];
                 this.jointValue[2] = values[2] !== null ? values[2] : this.jointValue[2];
-                this.position.copy(this.origPosition);
-                // Respect origin RPY when setting position
-                _tempAxis.set(1, 0, 0).applyEuler(this.rotation);
-                this.position.addScaledVector(_tempAxis, this.jointValue[0]);
-                _tempAxis.set(0, 1, 0).applyEuler(this.rotation);
-                this.position.addScaledVector(_tempAxis, this.jointValue[1]);
-                _tempAxis.set(0, 0, 1).applyEuler(this.rotation);
-                this.position.addScaledVector(_tempAxis, this.jointValue[2]);
-
                 this.jointValue[3] = values[3] !== null ? values[3] : this.jointValue[3];
                 this.jointValue[4] = values[4] !== null ? values[4] : this.jointValue[4];
                 this.jointValue[5] = values[5] !== null ? values[5] : this.jointValue[5];
-                this.quaternion.setFromEuler(
-                    _tempEuler.set(
-                        this.jointValue[3],
-                        this.jointValue[4],
-                        this.jointValue[5],
-                        'XYZ',
-                    ),
-                ).premultiply(this.origQuaternion);
+
+                // Compose transform of joint origin and transform due to joint values
+                _tempOrigTransform.compose(this.origPosition, this.origQuaternion, _tempScale);
+                _tempQuat.setFromEuler(
+                  _tempEuler.set(
+                    this.jointValue[3],
+                    this.jointValue[4],
+                    this.jointValue[5],
+                    'XYZ',
+                  ),
+                )
+                _tempPosition.set(this.jointValue[0], this.jointValue[1], this.jointValue[2]);
+                _tempTransform.compose(_tempPosition, _tempQuat, _tempScale);
+
+                // Calcualte new transform
+                const transformNew = _tempOrigTransform.premultiply(_tempTransform);
+                this.position.setFromMatrixPosition(transformNew)
+                this.rotation.setFromRotationMatrix(transformNew)
```

Comment:
`.premultiply` modifies in-place, and we don't want to allocate any new memory here by declaring a variable (in case it's being run at a very high rate by an unthrottled joint_states listener) .
```suggestion
                _tempOrigTransform.premultiply(_tempTransform);
                this.position.setFromMatrixPosition(_tempOrigTransform);
                this.rotation.setFromRotationMatrix(_tempOrigTransform);
```